File: backend/routes/trades.py
"""Trades + Portfolio routes."""
from typing import Optional
from fastapi import APIRouter, HTTPException, Header
from ..core import (
    db, new_id, now_iso, serialize, serialize_list,
    PlaceTradeBody,
)

router = APIRouter(tags=["trades"])


@router.post("/trades")
async def place_trade(body: PlaceTradeBody, x_wallet: Optional[str] = Header(None)):
    if not x_wallet:
        x_wallet = f"demo_{new_id()[:8]}.sol"
    market = await db.markets.find_one({"id": body.marketId})
    if not market:
        raise HTTPException(404, "Market not found")

    side = body.side.upper()
    price = market["yesPrice"] if side == "YES" else market["noPrice"]
    cost = round(body.shares * price, 2)
    fee = round(cost * 0.01, 4)

    trade = {
        "id": new_id(),
        "marketId": body.marketId,
        "userId": x_wallet,
        "wallet": x_wallet,
        "handle": None,
        "isAgent": False,
        "side": side,
        "kind": body.kind or "market",
        "shares": body.shares,
        "price": round(price, 4),
        "cost": cost,
        "fee": fee,
        "txSig": body.txSig or f"sig_{new_id()[:12]}",
        "createdAt": now_iso(),
    }
    await db.trades.insert_one(trade.copy())

    impact = min(0.05, body.shares / max(1, market["liquidity"]) * 0.5)
    new_yes_price = market["yesPrice"]
    new_yes_price = min(0.99, new_yes_price + impact) if side == "YES" else max(0.01, new_yes_price - impact)
    await db.markets.update_one({"id": body.marketId}, {
        "$set": {"yesPrice": round(new_yes_price, 4), "noPrice": round(1 - new_yes_price, 4)},
        "$inc": {"volume": cost, "participants": 1},
    })
    await db.price_points.insert_one({
        "id": new_id(), "marketId": body.marketId,
        "yesPrice": round(new_yes_price, 4), "noPrice": round(1 - new_yes_price, 4),
        "ts": now_iso(),
    })

    existing = await db.positions.find_one({"userId": x_wallet, "marketId": body.marketId})
    if existing:
        if side == "YES":
            shares = existing["yesShares"] + body.shares
            avg = (existing["avgYesCost"] * existing["yesShares"] + price * body.shares) / shares
            await db.positions.update_one({"id": existing["id"]}, {"$set": {"yesShares": shares, "avgYesCost": round(avg, 4)}})
        else:
            shares = existing["noShares"] + body.shares
            avg = (existing["avgNoCost"] * existing["noShares"] + price * body.shares) / shares
            await db.positions.update_one({"id": existing["id"]}, {"$set": {"noShares": shares, "avgNoCost": round(avg, 4)}})
    else:
        await db.positions.insert_one({
            "id": new_id(), "userId": x_wallet, "marketId": body.marketId,
            "yesShares": body.shares if side == "YES" else 0,
            "noShares": body.shares if side == "NO" else 0,
            "avgYesCost": round(price, 4) if side == "YES" else 0,
            "avgNoCost": round(price, 4) if side == "NO" else 0,
            "realizedPnl": 0.0, "createdAt": now_iso(),
        })

    return {"trade": serialize(trade)}


@router.get("/trades/recent/{market_id}")
async def recent_trades(market_id: str):
    trades = await db.trades.find({"marketId": market_id}).sort([("createdAt", -1)]).limit(30).to_list(30)
    return {"trades": serialize_list(trades)}


@router.get("/portfolio")
async def get_portfolio(x_wallet: Optional[str] = Header(None)):
    if not x_wallet:
        return {"summary": {"openValue": 0, "unrealized": 0, "realized": 0, "positions": 0}, "positions": [], "trades": []}
    positions = await db.positions.find({"userId": x_wallet}).to_list(100)
    trades = await db.trades.find({"userId": x_wallet}).sort([("createdAt", -1)]).limit(50).to_list(50)

    enriched_positions = []
    open_value = unrealized = realized = 0.0
    for pos in positions:
        market = await db.markets.find_one({"id": pos["marketId"]})
        if not market:
            continue
        p = serialize(pos)
        p["market"] = serialize(market)
        enriched_positions.append(p)
        if pos["yesShares"] > 0:
            v = pos["yesShares"] * market["yesPrice"]
            open_value += v
            unrealized += v - pos["yesShares"] * pos["avgYesCost"]
        if pos["noShares"] > 0:
            v = pos["noShares"] * market["noPrice"]
            open_value += v
            unrealized += v - pos["noShares"] * pos["avgNoCost"]
        realized += pos.get("realizedPnl", 0)

    enriched_trades = []
    for t in trades:
        market = await db.markets.find_one({"id": t["marketId"]})
        td = serialize(t)
        if market:
            td["market"] = {"question": market["question"], "category": market["category"]}
        enriched_trades.append(td)

    return {
        "summary": {
            "openValue": round(open_value, 2),
            "unrealized": round(unrealized, 2),
            "realized": round(realized, 2),
            "positions": len(enriched_positions),
        },
        "positions": enriched_positions,
        "trades": enriched_trades,
    }

File: backend/src/routes/trades.ts
import express, { Request, Response } from 'express';
import { prisma } from '../index';
import { newId } from '../utils/helpers';

const router = express.Router();

interface PlaceTradeBody {
  marketId: string;
  side: 'YES' | 'NO';
  shares: number;
  kind?: 'market' | 'limit';
  txSig?: string;
}

// Place a trade (buy YES/NO shares and update prices)
router.post('/', async (req: Request, res: Response): Promise<void> => {
  try {
    const body = req.body as PlaceTradeBody;
    const xWallet = (req.headers['x-wallet'] as string) || `demo_${newId().slice(0, 8)}.sol`;

    const market = await prisma.market.findUnique({
      where: { id: body.marketId },
    });

    if (!market) {
      res.status(404).json({ error: 'Market not found' });
      return;
    }

    if (!body.shares || body.shares <= 0) {
      res.status(400).json({ error: 'shares must be greater than 0' });
      return;
    }

    if (!['YES', 'NO'].includes(body.side.toUpperCase())) {
      res.status(400).json({ error: 'side must be YES or NO' });
      return;
    }

    const side = body.side.toUpperCase();
    const price = side === 'YES' ? market.yesPrice : market.noPrice;
    const cost = parseFloat((body.shares * price).toFixed(2));
    const fee = parseFloat((cost * 0.01).toFixed(4));

    // Create trade record
    const trade = await prisma.trade.create({
      data: {
        id: newId(),
        marketId: body.marketId,
        userId: xWallet,
        wallet: xWallet,
        isAgent: false,
        side,
        kind: body.kind || 'market',
        shares: body.shares,
        price: parseFloat(price.toFixed(4)),
        cost,
        fee,
        txSig: body.txSig || `sig_${newId().slice(0, 12)}`,
      },
    });

    // Calculate price impact
    const impact = Math.min(0.05, (body.shares / Math.max(1, market.liquidity)) * 0.5);
    let newYesPrice = market.yesPrice;
    if (side === 'YES') {
      newYesPrice = Math.min(0.99, newYesPrice + impact);
    } else {
      newYesPrice = Math.max(0.01, newYesPrice - impact);
    }
    const newNoPrice = 1 - newYesPrice;

    // Update market prices and stats
    await prisma.market.update({
      where: { id: body.marketId },
      data: {
        yesPrice: parseFloat(newYesPrice.toFixed(4)),
        noPrice: parseFloat(newNoPrice.toFixed(4)),
        volume: market.volume + cost,
        participants: market.participants + 1,
      },
    });

    // Record price point
    await prisma.pricePoint.create({
      data: {
        id: newId(),
        marketId: body.marketId,
        yesPrice: parseFloat(newYesPrice.toFixed(4)),
        noPrice: parseFloat(newNoPrice.toFixed(4)),
      },
    });

    // Update or create position
    const existingPosition = await prisma.position.findUnique({
      where: {
        marketId_userId: {
          marketId: body.marketId,
          userId: xWallet,
        },
      },
    });

    if (existingPosition) {
      if (side === 'YES') {
        const newShares = existingPosition.yesShares + body.shares;
        const avgCost = parseFloat(
          (
            (existingPosition.avgYesCost * existingPosition.yesShares + price * body.shares) /
            newShares
          ).toFixed(4)
        );
        await prisma.position.update({
          where: { id: existingPosition.id },
          data: {
            yesShares: newShares,
            avgYesCost: avgCost,
          },
        });
      } else {
        const newShares = existingPosition.noShares + body.shares;
        const avgCost = parseFloat(
          (
            (existingPosition.avgNoCost * existingPosition.noShares + price * body.shares) /
            newShares
          ).toFixed(4)
        );
        await prisma.position.update({
          where: { id: existingPosition.id },
          data: {
            noShares: newShares,
            avgNoCost: avgCost,
          },
        });
      }
    } else {
      await prisma.position.create({
        data: {
          id: newId(),
          marketId: body.marketId,
          userId: xWallet,
          yesShares: side === 'YES' ? body.shares : 0,
          noShares: side === 'NO' ? body.shares : 0,
          avgYesCost: side === 'YES' ? parseFloat(price.toFixed(4)) : 0,
          avgNoCost: side === 'NO' ? parseFloat(price.toFixed(4)) : 0,
        },
      });
    }

    res.status(201).json({ trade });
    return;
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to place trade' });
    return;
  }
});

// Get recent trades for a market
router.get('/market/:marketId', async (req: Request, res: Response) => {
  try {
    const trades = await prisma.trade.findMany({
      where: { marketId: req.params.marketId },
      orderBy: { createdAt: 'desc' },
      take: 30,
    });

    res.json({ trades });
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch trades' });
  }
});

// Get portfolio for user
router.get('/portfolio', async (req: Request, res: Response): Promise<void> => {
  try {
    const xWallet = (req.headers['x-wallet'] as string) || `demo_${newId().slice(0, 8)}.sol`;

    if (!xWallet || xWallet.startsWith('demo_')) {
      res.json({
        summary: { openValue: 0, unrealized: 0, realized: 0, positions: 0 },
        positions: [],
        trades: [],
      });
      return;
    }

    const positions = await prisma.position.findMany({
      where: { userId: xWallet },
      include: { market: true },
    });

    const trades = await prisma.trade.findMany({
      where: { userId: xWallet },
      orderBy: { createdAt: 'desc' },
      take: 50,
    });

    let openValue = 0;
    let unrealized = 0;
    const realized = positions.reduce((sum, p) => sum + p.realizedPnl, 0);

    const enrichedPositions = positions.map((pos) => {
      if (pos.yesShares > 0) {
        const v = pos.yesShares * pos.market.yesPrice;
        openValue += v;
        unrealized += v - pos.yesShares * pos.avgYesCost;
      }
      if (pos.noShares > 0) {
        const v = pos.noShares * pos.market.noPrice;
        openValue += v;
        unrealized += v - pos.noShares * pos.avgNoCost;
      }
      return {
        ...pos,
        market: {
          question: pos.market.question,
          category: pos.market.category,
        },
      };
    });

    const enrichedTrades = trades.map((t) => ({
      ...t,
      market: { question: 'N/A', category: 'N/A' },
    }));

    res.json({
      summary: {
        openValue: parseFloat(openValue.toFixed(2)),
        unrealized: parseFloat(unrealized.toFixed(2)),
        realized: parseFloat(realized.toFixed(2)),
        positions: enrichedPositions.length,
      },
      positions: enrichedPositions,
      trades: enrichedTrades,
    });
    return;
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch portfolio' });
    return;
  }
});

export default router;
